### DISCUSSION TITLE
Testing component which use useEffect

### DISCUSSION THREAD

**Discussion by A11yship** [neuefische/questions#133] [Web] [answered]
Hello,
sine I added a useEffect Hook in my form component the test fails. You can see it in my [PR](https://github.com/A11yship/capstone-project/pull/29). If I comment off the lines 16-19 in the `Form.js` the form test passes.
Do I have to mock the useEffect? If yes, how do I do that?

Thanks for helping me.

**Comment by alexclaes** [comment, marked as answer]
Using a component's prop as a dependency for an effect to trigger a state change is kind of an anti-pattern and messes with the render cycle.

It would work in the browser, but breaks the test. If you place a `console.log` within the effect callback function, you can see the effect gets called all the time.

- Typing in the input field changes the sate
- Changed the state triggers re-rendering of the `<Form>` component
- The prop `task` is evaluated with a new(!) object as default value
- This triggers the effect again, which triggers the state change again

I'm sure you don't need the `useEffect` at all. Using the `task` prop to set an initial state for `taskName` and `duration` is sufficient.

However, this will introduce a bug when a user changes from the edit-view to the create-view (becaus both use the same component instance) I highly recommend to use separated Next.js pages for the edit-view and th create-view. By using separated pages you can enforce separated component instances.

**Reply by A11yship** [reply]
Thanks for the recommendation with the two pages. I have done that and delete the `useEffect`. Now evverything works fine. I didn't saw this solution for the bug you mentioned.
